kreuzberg/_ocr/_table_extractor.py

```diff
@@ -0,0 +1,404 @@
+"""Table extraction from Tesseract TSV output."""
+
+from __future__ import annotations
+
+import csv
+from dataclasses import dataclass
+from io import StringIO
+from typing import TYPE_CHECKING, Protocol, runtime_checkable
+
+import numpy as np
+
+from kreuzberg.exceptions import ParsingError
+
+if TYPE_CHECKING:
+    from kreuzberg._types import TSVWord
+
+try:
+    from scipy.cluster.hierarchy import fclusterdata
+
+    SCIPY_AVAILABLE = True
+except ImportError:
+    SCIPY_AVAILABLE = False
+
+
+@runtime_checkable
+class TableExtractor(Protocol):
+    """Protocol for table extraction algorithms."""
+
+    def extract_words(self, tsv_data: str) -> list[TSVWord]:
+        """Parse TSV data into word objects.
+
+        Args:
+            tsv_data: Raw TSV output from Tesseract.
+
+        Returns:
+            List of word dictionaries with position and text data.
+        """
+        ...
+
+    def detect_columns(self, words: list[TSVWord]) -> list[int]:
+        """Detect column positions using clustering.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            Sorted list of column X positions.
+        """
+        ...
+
+    def detect_rows(self, words: list[TSVWord]) -> list[int]:
+        """Detect row positions using clustering.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            Sorted list of row Y positions.
+        """
+        ...
+
+    def reconstruct_table(self, words: list[TSVWord]) -> list[list[str]]:
+        """Reconstruct table as 2D array.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            2D list representing the table structure.
+        """
+        ...
+
+    def to_markdown(self, table: list[list[str]]) -> str:
+        """Convert table to markdown format.
+
+        Args:
+            table: 2D list representing the table.
+
+        Returns:
+            Markdown-formatted table string.
+        """
+        ...
+
+
+@dataclass
+class TesseractTableExtractor:
```

Comment:
this does not need to be a class, let alone a dataclass. Remove the protocol, then change the logic here to be functions. Use kwargs for the functions (not args)